scripts/config.py

```diff
@@ -0,0 +1,408 @@
+      set, raise `KeyError` (even if a definition for `name` is present
+      but commented out).
+    * `config[name] = value` sets the value associated to `name`. `name`
+      must be known, but does not need to be set. This does not cause
+      name to become set.
+    """
+
+    def __init__(self):
+        self.settings = {}
+
+    def __contains__(self, name):
+        """True if the given symbol is active (i.e. set).
+
+        False if the given symbol is not set, even if a definition
+        is present but commented out.
+        """
+        return name in self.settings and self.settings[name].active
+
+    def all(self, *names):
+        """True if all the elements of names are active (i.e. set)."""
+        return all(self.__contains__(name) for name in names)
+
+    def any(self, *names):
+        """True if at least one symbol in names are active (i.e. set)."""
+        return any(self.__contains__(name) for name in names)
+
+    def known(self, name):
+        """True if a #define for name is present, whether it's commented out or not."""
+        return name in self.settings
+
+    def __getitem__(self, name):
+        """Get the value of name, i.e. what the preprocessor symbol expands to.
+
+        If name is not known, raise KeyError. name does not need to be active.
+        """
+        return self.settings[name].value
+
+    def get(self, name, default=None):
+        """Get the value of name. If name is inactive (not set), return default.
+
+        If a #define for name is present and not commented out, return
+        its expansion, even if this is the empty string.
+
+        If a #define for name is present but commented out, return default.
+        """
+        if name in self.settings:
+            return self.settings[name].value
+        else:
+            return default
+
+    def __setitem__(self, name, value):
+        """If name is known, set its value.
+
+        If name is not known, raise KeyError.
+        """
+        self.settings[name].value = value
+
+    def set(self, name, value=None):
+        """Set name to the given value and make it active.
+
+        If value is None and name is already known, don't change its value.
+        If value is None and name is not known, set its value to the empty
+        string.
+        """
+        if name in self.settings:
+            if value is not None:
+                self.settings[name].value = value
+            self.settings[name].active = True
+        else:
+            self.settings[name] = Setting(True, name, value=value)
+
+    def unset(self, name):
+        """Make name unset (inactive).
+
+        name remains known.
+        """
+        self.set(name)
+        self.settings[name].active = False
+
+    def adapt(self, adapter):
+        """Run adapter on each known symbol and (de)activate it accordingly.
+
+        `adapter` must be a function that returns a boolean. It is called as
```

Comment:
I'm not familiar with any “common meaning of the word”. To me, this method adapts the settings. I'm willing to change the name, but I don't find your proposals clearer.

Reply:
How about `transform(transformation)`?

Reply:
The "adapter" may be associated with one of the "design patterns", where it is a technique of piling APIs on top of each other. If you think of physical adapters for plugs and cables, one would expect some data flow, e.g. a conversion function. The "transform" you propose implies data mutation.
I don't think the functions we discuss perform any of that.

They evaluate some state (here, a config setting) and return some Boolean value. To me that is a predicate, and I don't know any better word for that. So, if you don't like "predicate", please use whichever word you find best; I'm OK with the original "adapter" as well.

Reply:
Now I see that for a function that returns a boolean, `transformation` and `adapter` aren't good names. I'll keep `adapt` but change the argument to `predicate`.

Reply:
On second thoughts, no, I've identified what bothers me with `predicate`: a predicate studies an object and returns a boolean. This function studies an object and _transforms_ a boolean into another. It doesn't just say “activate this symbol” or “deactivate this symbol” based on the symbol: it can also say “keep this symbol in its current state”. So I'll stick with `adapt(adapter)` unless a better name comes up.

Reply:
On second thoughts, no, I've identified what bothers me with `predicate`. The problem struck me when I renamed `baremetal_adapter` and friends. A predicate studies an object and returns a boolean. This function studies an object and _transforms_ a boolean into another. It doesn't just say “activate this symbol” or “deactivate this symbol” based on the symbol: it can also say “keep this symbol in its current state”. So I'll stick with `adapt(adapter)` unless a better name comes up.